File: ljmd-c/src/mpi/main_mpi.c
/* 
 * simple lennard-jones potential MD code with velocity verlet.
 * units: Length=Angstrom, Mass=amu; Energy=kcal
 *
 * baseline c version.
 */

#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>
#include <math.h>
#include "data.h"
#include "prototypes.h"

/* main */
int main(int argc, char **argv) 
{
    int nprint, i;
    char restfile[BLEN], trajfile[BLEN], ergfile[BLEN], line[BLEN];
    FILE *fp,*traj,*erg;
    mdsys_t sys;

    sys.mpicomm=MPI_COMM_WORLD;

    MPI_Init(&argc,&argv);
    MPI_Comm_rank(sys.mpicomm,&sys.mpirank);
    MPI_Comm_size(sys.mpicomm,&sys.nprocs);

    /* read input file */
    /*if (sys.mpirank==0){
    	if(get_a_line(stdin,line)) return 1;
    	sys.natoms=atoi(line);
    	if(get_a_line(stdin,line)) return 1;
    	sys.mass=atof(line);
    	
	if(get_a_line(stdin,line)) return 1;
    	sys.epsilon=atof(line);

    	
    	if(get_a_line(stdin,line)) return 1;
    	sys.sigma=atof(line);
    	
    	
	if(get_a_line(stdin,line)) return 1;
    	sys.rcut=atof(line);
    	
    	
	if(get_a_line(stdin,line)) return 1;
    	sys.box=atof(line);
    	

	if(get_a_line(stdin,restfile)) return 1;
    	if(get_a_line(stdin,trajfile)) return 1;
    	if(get_a_line(stdin,ergfile)) return 1;
    	if(get_a_line(stdin,line)) return 1;
    	sys.nsteps=atoi(line);
	if(get_a_line(stdin,line)) return 1;
    	sys.dt=atof(line);	
    	
	if(get_a_line(stdin,line)) return 1;
    	nprint=atoi(line);

    	
    }

	MPI_Bcast(&sys.dt,1,MPI_DOUBLE,0,sys.mpicomm);
	MPI_Bcast(&sys.nsteps,1,MPI_DOUBLE,0,sys.mpicomm);
	MPI_Bcast(&sys.box,1,MPI_DOUBLE,0,sys.mpicomm);
	MPI_Bcast(&sys.rcut,1,MPI_DOUBLE,0,sys.mpicomm);
	MPI_Bcast(&sys.natoms,1,MPI_DOUBLE,0,sys.mpicomm); 
	MPI_Bcast(&sys.sigma,1,MPI_DOUBLE,0,sys.mpicomm);
	MPI_Bcast(&sys.epsilon,1,MPI_DOUBLE,0,sys.mpicomm);
	MPI_Bcast(&sys.mass,1,MPI_DOUBLE,0,sys.mpicomm);
    */

	sys.natoms=108;
	sys.mass=39.948;
	sys.epsilon=0.2379;
	sys.sigma=3.405;
	sys.rcut=8.5;
	sys.box=17.1580;
	sys.dt=5.0;
	sys.nsteps=1000;	

	sprintf(restfile,"argon_108.rest");
	
	sys.nsize=sys.natoms/sys.nprocs;
 
    /* allocate memory */
    sys.rx=(double *)malloc(sys.natoms*sizeof(double));
    sys.ry=(double *)malloc(sys.natoms*sizeof(double));
    sys.rz=(double *)malloc(sys.natoms*sizeof(double));
    sys.vx=(double *)malloc(sys.natoms*sizeof(double));
    sys.vy=(double *)malloc(sys.natoms*sizeof(double));
    sys.vz=(double *)malloc(sys.natoms*sizeof(double));
   
    // Buffer for Broadcast
    sys.cx=(double *)malloc(sys.natoms*sizeof(double));
    sys.cy=(double *)malloc(sys.natoms*sizeof(double));
    sys.cz=(double *)malloc(sys.natoms*sizeof(double));
	if (sys.mpirank==0){
    		sys.fx=(double *)malloc(sys.natoms*sizeof(double));
    		sys.fy=(double *)malloc(sys.natoms*sizeof(double));
    		sys.fz=(double *)malloc(sys.natoms*sizeof(double));
	}


    /* read restart */
    if (sys.mpirank==0){
    	fp=fopen(restfile,"r");
    	if(fp) {
        	for (i=0; i<sys.natoms; ++i) {
            		fscanf(fp,"%lf%lf%lf",sys.rx+i, sys.ry+i, sys.rz+i);
        	}
        	for (i=0; i<sys.natoms; ++i) {
            		fscanf(fp,"%lf%lf%lf",sys.vx+i, sys.vy+i, sys.vz+i);
        	}
        	fclose(fp);
        	azzero(sys.fx, sys.natoms);
        	azzero(sys.fy, sys.natoms);
        	azzero(sys.fz, sys.natoms);
    	} else {
        	perror("cannot read restart file");
        	return 3;
    	}
    }
	MPI_Barrier(sys.mpicomm);
    /* initialize forces and energies.*/
    sys.nfi=0;
    force(&sys);

if (sys.mpirank==0){
	for (int i=0;i<sys.natoms;i++){
		printf("Force on atom %d: (%f,%f,%f)\n",i,sys.fx[i],sys.fy[i],sys.fz[i]);
	}
	printf("\n");
}
    //ekin(&sys);
   
   /*if (sys.mpirank==0){ 
    erg=fopen(ergfile,"w");
    traj=fopen(trajfile,"w");

    printf("Starting simulation with %d atoms for %d steps.\n",sys.natoms, sys.nsteps);
    printf("     NFI            TEMP            EKIN                 EPOT              ETOT\n");
    output(&sys, erg, traj);
   }*/

    /**************************************************/
    /* main MD loop */
   // for(sys.nfi=1; sys.nfi <= sys.nsteps; ++sys.nfi) {

   //     /* write output, if requested */
   //     if ((sys.nfi % nprint) == 0 && sys.mpirank==0)
   //         output(&sys, erg, traj);

   //     /* propagate system and recompute energies */
   //     velverlet(&sys);
   //     ekin(&sys);
    //}
    /**************************************************/

    /* clean up: close files, free memory */
    printf("Simulation Done.\n");
    //fclose(erg);
    //fclose(traj);

    /*free(sys.rx);
    free(sys.ry);
    free(sys.rz);
    free(sys.vx);
    free(sys.vy);
    free(sys.vz);
    free(sys.fx);
    free(sys.fy);
    free(sys.fz);*/

    MPI_Finalize();
    return 0;
}

File: ljmd-c/src_py/main_mpi.py
import sys
from data import *
from utilities import *
from output import *
from energy import *
import time
from mpi4py import MPI

def create_system(S, input_contents):
   for count,line in enumerate(input_contents):
       (R,value)=get_a_line(line)
       if(R==0):
          if(  count== 0): S.natoms  = c_int(int(value))
          elif(count== 1): S.mass    = c_double(float(value))
          elif(count== 2): S.epsilon = c_double(float(value))
          elif(count== 3): S.sigma   = c_double(float(value))
          elif(count== 4): S.rcut    = c_double(float(value))
          elif(count== 5): S.box     = c_double(float(value))
          elif(count== 6): restfile  = value
          elif(count== 7): trajfile  = value
          elif(count== 8): ergfile   = value
          elif(count== 9): S.nsteps  = c_int(int(value))
          elif(count==10): S.dt      = c_double(float(value))
          elif(count==11): nprint    = int(value)
   return restfile,trajfile,ergfile,nprint

#****************************************************************************************

# MAIN

system = mdsys_t()

comm = MPI.COMM_WORLD
comm_ptr = MPI._addressof(comm)
comm_val = MPI_Comm.from_address(comm_ptr)

system.mpicomm = comm_val
system.mpirank  = comm.Get_rank()
system.nprocs = comm.Get_size()

if (system.mpirank == 0):
    
    # read file from stdin

    restfile,trajfile,ergfile,nprint=create_system(system, sys.stdin)
    
    # otherwise read input file directly

    #f = open("../examples/argon_2916.inp", "r")
    #restfile,trajfile,ergfile,nprint=create_system(system, f) 
    #f.close()
    
system.dt = comm.bcast(system.dt, root = 0)
system.nsteps = comm.bcast(system.nsteps, root = 0)
system.box = comm.bcast(system.box, root = 0)
system.rcut = comm.bcast(system.rcut, root = 0)
system.natoms = comm.bcast(system.natoms, root = 0) 
system.sigma = comm.bcast(system.sigma, root = 0)
system.epsilon = comm.bcast(system.epsilon, root = 0)
system.mass = comm.bcast(system.mass, root = 0)

system.nthreads = 1

# allocate memory

system.rx = (c_double * system.natoms)()
system.ry = (c_double * system.natoms)()
system.rz = (c_double * system.natoms)()
system.vx = (c_double * system.natoms)()
system.vy = (c_double * system.natoms)()
system.vz = (c_double * system.natoms)()

system.cx = (c_double * system.natoms)()
system.cy = (c_double * system.natoms)()
system.cz = (c_double * system.natoms)()

if system.mpirank == 0:
    system.fx = (c_double * system.natoms)()
    system.fy = (c_double * system.natoms)()
    system.fz = (c_double * system.natoms)()

# read restart

if system.mpirank == 0:

    fp = open("../examples/" + restfile, "r")

    for i in range(system.natoms):
        rx, ry, rz = fp.readline().split()
        system.rx[i] = c_double(float(rx))
        system.ry[i] = c_double(float(ry))
        system.rz[i] = c_double(float(rz))
    for i in range(system.natoms):
        vx, vy, vz = fp.readline().split()
        system.vx[i] = c_double(float(vx))
        system.vy[i] = c_double(float(vy))
        system.vz[i] = c_double(float(vz))

    fp.close()
    
comm.barrier()

# initialize forces and energies

system.nfi = 0

fso = CDLL("../Obj-mpi/libforce_mpi.so" )
fso.force.argtypes =[POINTER(mdsys_t)] #Structure

vso = CDLL("../Obj-mpi/libvelverlet.so" )
vso.velverlet.argtypes =[POINTER(mdsys_t)] #Structure
vso.velverlet_first.argtypes =[POINTER(mdsys_t)] #Structure
vso.velverlet_second.argtypes =[POINTER(mdsys_t)] #Structure

fso.force(system)

if system.mpirank == 0:
    ekin(system)

    erg = open(ergfile, "w")
    traj = open(trajfile, "w")

    print("Starting simulation with %d atoms for %d steps." % (system.natoms, system.nsteps));
    print("     NFI            TEMP            EKIN                 EPOT              ETOT");

    output(system, erg, traj)

#****************************************************************************************

# main MD loop with timing
 
t = 0.0
final_t = 0.0

for system.nfi in range(1, system.nsteps + 1):

    # write output, if requested 
    if (system.mpirank == 0 and (system.nfi % nprint) == 0):
        output(system, erg, traj);
    
    t_tmp = time.time()
    
    # propagate system and recompute energies 
    if system.mpirank == 0:
        vso.velverlet_first(system);
    
    fso.force(system)
    
    if system.mpirank == 0:
        vso.velverlet_second(system);
        ekin(system)
    
    t += time.time() - t_tmp

final_t = comm.reduce(t, op = MPI.MAX, root = 0)

print("Time proc %d: %.6f" %(system.mpirank, t))
if system.mpirank == 0:
    print("Evolve time: %.6f" %(final_t))
    
#****************************************************************************************

# clean up: close files

if system.mpirank == 0:
    print("Simulation Done.");

    erg.close()
    traj.close()

#****************************************************************************************
